jax/interpreters/pxla.py

```diff
@@ -613,35 +613,19 @@ def parallel_callable(fun, backend, axis_name, axis_size, global_axis_size,
     raise ValueError("'devices' argument to pmap must be non-empty, or None.")
 
   # Determine global_axis_size for use in AxisEnv.
+  must_run_on_all_devices = True
   if devices:
     assert global_axis_size is None  # Checked in api.py
     global_axis_size = len(devices)
-  elif xb.host_count() > 1:
-    if global_axis_size is None:
-      # TODO(skye): relax this constraint or provide functionality for
-      # automatically passing appropriate `devices`.
-      # TODO(trevorcai): This check forces us to provide global_axis_size for
-      # all pmaps on pmap-on-pod. Can we do it after tracing?
-      if axis_size != xb.local_device_count():
-        raise ValueError(
-            "On multi-host platforms, the input to pmapped functions must have "
-            "leading axis size equal to the number of local devices if no "
-            "`devices` argument is specified. Got axis_size=%d, "
-            "num_local_devices=%d" % (axis_size, xb.local_device_count()))
-      global_axis_size = xb.device_count()
+    must_run_on_all_devices = False
   else:
-    if global_axis_size is not None:
-      if global_axis_size != axis_size:
-        raise ValueError(
-            "Specified axis_size {} doesn't match received axis_size {}.".format(
-                global_axis_size, axis_size))
-    else:
-      global_axis_size = axis_size
-
-  log_priority = logging.WARNING if FLAGS.jax_log_compiles else logging.DEBUG
-  logging.log(log_priority,
-              "Compiling {} for {} devices with args {}.".format(
-                  fun.__name__, global_axis_size, avals))
+    expected = axis_size * xb.host_count()
```

Comment:
I think this is problematic when evaluating this logic on the pmap of nested cmap/pmap.

Consider evaluating the following on 2 hosts of 8 devices each.
```jax.pmap(jax.pmap(lambda x: x * 2, global_axis_size=4), global_axis_size=4)```
The inner, local, pmap axis will computed `expected = 4 * 2 = 8`, which does not match the specified `global_axis_size`.

The original (convoluted) control flow only infers`global_axis_size` from data & context when it's `None` for this reason. (Alternatively: examine the current AxisEnv and don't do this when inside a pmap.)

Is this logic problematic in the context of sharding?

Reply:
Blah you're right. I forgot this code is run for inner pmaps, so I was only considering the outermost case. I put back most of the convoluted original logic, and also check the DynamicAxisEnv to only use `axis_size * host_count` for the outermost pmap if there's no user-provided `global_axis_env`.

I also made it an error if no `global_axis_size` is specified for an inner pmap. This should be an error right?

This works for sharding, because inner sharded_jits go through a simpler non-multihost-aware path.

(BTW, the only reason this runs for inner pmaps is to do partial evaluation, i.e. to evaluate any non-data-dependent values in the inner pmap, which is gonna go away soon. Then this code will only run for the outermost pmap, which will be much simpler to reason about.)

Reply:
This seems like a reasonable fix.
Is this yet another problem that omnistaging is going to solve? Y'all are making it sound like the greatest thing since sliced bread :)

Reply:
Yes, yes it is!